src/common/dav/tx.c

```diff
@@ -545,26 +563,7 @@ dav_tx_begin(dav_obj_t *pop, jmp_buf env, ...)
 
 	DAV_SLIST_INSERT_HEAD(&tx->tx_entries, txd, tx_entry);
 
-	if (tx->stage == DAV_TX_STAGE_NONE) {
-		struct umem_wal_tx *utx;
-
-		tx->stage = DAV_TX_STAGE_WORK;
-		utx = pop->do_utx;
-		if (utx == NULL) {
-			utx = dav_umem_wtx_new(pop);
-			if (utx == NULL) {
-				err = ENOMEM;
-				goto err_abort;
-			}
-		}
-		err = dav_wal_tx_reserve(pop);
-		if (err) {
-			D_ERROR("so_wal_reserv failed, "DF_RC"\n", DP_RC(err));
-			D_FREE(utx);
-			pop->do_utx = NULL;
-			goto err_abort;
-		}
-	}
+	tx->stage = DAV_TX_STAGE_WORK;
```

Comment:
We can't access tx after the context switch.

Reply:
Since wal_tx_reserve() guarantees first call first return, it should be fine, but I strongly suggest move this into the "else if (tx->stage == DAV_TX_STAGE_NONE)" branch and add a assert of "tx->stage == DAV_TX_STAGE_NONE" before change it.

Reply:
Similar assert is already present in line 543. We reach this point only if tx->stage is DAV_TX_STAGE_NONE or DAV_TX_STAGE_WORK.

Reply:
You mean the "FATAL("Invalid stage..)" above? That looks assert on different things to me, my proposal is like:

else if (tx->stage == DAV_TX_STAGE_NONE) {
    wal_reserve();
    D_ASSERT(tx->stage == DAV_TX_STAGE_NONE);  /* Nobody change tx stage while I yield in wal_reserve() */
    tx->stage = DAV_TX_STAGE_WORK;
}

I'm fine with cleanup it in follow-on PR.